pkg/admission/cert/writer/secret.go

```diff
@@ -0,0 +1,246 @@
+		if strings.HasPrefix(k, SecretCertProvisionAnnotationKeyPrefix) {
+			webhookName := strings.TrimPrefix(k, SecretCertProvisionAnnotationKeyPrefix)
+			secretWebhookMap[webhookName] = &webhookAndSecret{
+				secret: types.NewNamespacedNameFromString(v),
+			}
+		}
+	}
+}
+
+func (s *secretReadWriter) ensureCert() error {
+	for _, v := range s.webhookMap {
+		err := handleCommon(v.webhook, s)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// secretReadWriter deals with writing to the k8s secrets.
+type secretReadWriter struct {
+	client        client.Client
+	certGenerator generator.CertGenerator
+
+	webhookConfig runtime.Object
+	webhookMap    map[string]*webhookAndSecret
+}
+
+type webhookAndSecret struct {
+	webhook *admissionregistrationv1beta1.Webhook
+	secret  apitypes.NamespacedName
+}
+
+var _ certReadWriter = &secretReadWriter{}
+
+func (s *secretReadWriter) write(webhookName string) (
+	*generator.Artifacts, error) {
+	v := s.webhookMap[webhookName]
+
+	webhook := v.webhook
+	commonName, err := dnsNameForWebhook(&webhook.ClientConfig)
+	if err != nil {
+		return nil, err
+	}
+	certs, err := s.certGenerator.Generate(commonName)
+	if err != nil {
+		return nil, err
+	}
+	secret := certsToSecret(certs, v.secret)
+	// TODO: fix it: see the TODO in the method
+	err = setOwnerRef(secret, s.webhookConfig)
+	if err != nil {
+		return nil, err
+	}
+	err = s.client.Create(nil, secret)
+	return certs, err
+}
+
+func (s *secretReadWriter) overwrite(webhookName string) (
+	*generator.Artifacts, error) {
+	v := s.webhookMap[webhookName]
+
+	webhook := v.webhook
+	commonName, err := dnsNameForWebhook(&webhook.ClientConfig)
+	if err != nil {
+		return nil, err
+	}
+	certs, err := s.certGenerator.Generate(commonName)
+	if err != nil {
+		return nil, err
+	}
+	secret := certsToSecret(certs, v.secret)
+	// TODO: fix it: see the TODO in the method
+	err = setOwnerRef(secret, s.webhookConfig)
+	if err != nil {
+		return nil, err
+	}
+	err = s.client.Update(nil, secret)
+	return certs, err
+}
+
+func (s *secretReadWriter) read(webhookName string) (*generator.Artifacts, error) {
+	v := s.webhookMap[webhookName]
+	secret := &corev1.Secret{}
+	err := s.client.Get(nil, v.secret, secret)
+	return secretToCerts(secret), err
+}
+
+// setOwnerRef marks the webhook as the owner of the secret by setting the ownerReference in the secret.
+func setOwnerRef(secret, webhookConfig runtime.Object) error {
+	accessor, err := meta.Accessor(webhookConfig)
+	if err != nil {
+		return err
+	}
+	// TODO: typeAccessor.GetAPIVersion() and typeAccessor.GetKind() returns empty apiVersion and Kind, fix it.
+	typeAccessor, err := meta.TypeAccessor(webhookConfig)
+	if err != nil {
+		return err
+	}
+	blockOwnerDeletion := false
+	// Due to
+	// https://github.com/kubernetes/kubernetes/blob/5da925ad4fd070e687dc5255c177d5e7d542edd7/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/controller_ref.go#L35
+	isController := true
+	ownerRef := metav1.OwnerReference{
```

Comment:
Why not use this:

https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controllerutil/controllerutil.go#L31